[Zen2] Introduce gossip-like discovery of master nodes

DaveCTurner opened the pull request:
This commit introduces the `PeerFinder` which can be used to collect the
identities of the master-eligible nodes in a masterless cluster, based on the
`UnicastHostsProvider`, the nodes in the `ClusterState`, and nodes that other
nodes have discovered.

elasticmachine commented:
Pinging @elastic/es-distributed

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+
+public abstract class PeerFinder extends AbstractLifecycleComponent {
+
+    public static final String REQUEST_PEERS_ACTION_NAME = "internal:discovery/zen2/requestpeers";
```
Ok, done

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+
+    // the time between attempts to find all peers
+    public static final Setting<TimeValue> CONSENSUS_FIND_PEERS_DELAY_SETTING =
+        Setting.timeSetting("discovery.zen2.find_peers_delay",
```
Ok, done

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+    public static final String REQUEST_PEERS_ACTION_NAME = "internal:discovery/zen2/requestpeers";
+
+    // the time between attempts to find all peers
+    public static final Setting<TimeValue> CONSENSUS_FIND_PEERS_DELAY_SETTING =
```
Ok, done

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+        return getFoundPeersSet();
+    }
+
+    public boolean foundQuorumFrom(VotingConfiguration votingConfiguration) {
```
Ok, fixed

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+        this.executorServiceFactory = executorServiceFactory;
+    }
+
+    public Iterable<DiscoveryNode> getFoundPeers() {
```
Ok, fixed

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+            return;
+        }
+
+        // Must not hold lock when calling out to Legislator to avoid a lock cycle
```
Added a TODO - there are other comments about this code elsewhere.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+    protected void doClose() {
+    }
+
+    abstract DiscoveryNodes getLastAcceptedNodes();
```
Ok, done.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+
+    abstract DiscoveryNodes getLastAcceptedNodes();
+
+    abstract void onMasterFoundByProbe(DiscoveryNode masterNode, long term);
```
Ok, done.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+        }
+
+        private FoundPeer getFoundPeer(DiscoveryNode discoveryNode) {
+            // no synchronisation required: computeIfAbsent is atomic as foundPeers is a newConcurrentMap()
```
Ok, inlined.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+
+        // Must not hold lock when calling out to Legislator to avoid a lock cycle
+        assert holdsLock() == false : "Peerfinder mutex is held in error";
+        final DiscoveryNodes lastAcceptedNodes = getLastAcceptedNodes();
```
This is tricky because the wake-up has to get the nodes from somewhere. We could pass in a Supplier but it'd still need to do something with locking (maybe this can be avoided with a `LastAcceptedNodesResponder`)? I added a TODO for further discussion.

DaveCTurner replied:
Oh yes, good point. I pushed 2177d0a1a2a.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+            masterNodes.forEach(e -> startProbe(e.value));
+
+            if (resolveInProgress.compareAndSet(false, true)) {
+                executorService.get().execute(labelRunnable(() -> {
```
I had a brief look at Zen1 and it seemed that name resolution happens inline with the whole pinging process. Changed to use the generic threadpool anyway.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+                        transportService, resolveTimeout)));
+
+                    // localNode is excluded from buildDynamicNodes
+                    providedAddresses.add(transportService.getLocalNode().getAddress());
```
No, fixed.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+                        providedAddresses.forEach(ActivePeerFinder.this::startProbe);
+                    }
+
+                    resolveInProgress.set(false);
```
Ok, done.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeersResponse.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.Optional;
+
+public class PeersResponse extends TransportResponse {
+    private final Optional<DiscoveryNode> masterNode;
```
not sure I follow (no pun intended). A follower would return isActiveMaster = false + the active master as singleton in the list of masterEligibleNodes, which would lead to exact same behavior as what's currently in the PR if I understand this correctly.

DaveCTurner replied:
We discussed on Zoom. There's no strong argument either way. I slightly prefer that as it is now there's a difference in the responses between a follower and a candidate that knows of a single node, although we don't use this difference anywhere. I also see that moving to a single list (a singleton from followers) or a boolean would remove a couple of lines of code that deal with the master node as a special case. We'll leave it as is.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,453 @@
+                }
+
+                logger.trace("startProbe({}) found disconnected {}, probing again", transportAddress, cachedNode);
+                connectedNodes.remove(transportAddress, cachedNode);
```
`foundPeers` is no longer a thing - we just use `connectedNodes` throughout.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,496 @@
+
+            if (inFlightProbes.add(transportAddress)) {
+                logger.trace("startProbe({}) acquired probeLock", transportAddress);
+                executorService.get().execute(new AbstractRunnable() {
```
There are two levels of async here that could arguably collapse to one, with `transportAddressConnector` doing its magic with its own `executorService` (possibly the same one).

DaveCTurner commented:
On reflection, I think that collecting together the state attached to each `TransportAddress` would be a nicer structure for this based on how it looks now. Don't review in detail, I will pursue that idea tomorrow.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/coordination/PeerFinderTests.java:
```diff
@@ -0,0 +1,671 @@
+                                listener.onResponse(discoveryNode);
+                                return;
+                            } else {
+                                listener.onFailure(new ElasticsearchException("non-master node " + discoveryNode)); // TODO better exception
```
Does this need a better exception or is this ok?

ywelsch replied:
for the tests, I think it doesn't matter

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,484 @@
+    private final TimeValue findPeersDelay;
+    private final TimeValue resolveTimeout;
+
+    private final Object mutex = new Object();
```
The new structure suggests that perhaps we don't need such coarse locking, but I do not think this is worth pursuing until we have some evidence that we need something finer. This is not very high-throughput code.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,484 @@
+            }
+        }
+
+        return onPeersRequestWhenInactive(peersRequest.getSourceNode());
```
Ok, done in 9356a26980a300c5fee4ebce4b6609b7bd2b2f57

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,484 @@
+    }
+
+    public void setCurrentTerm(long currentTerm) {
+        // Volatile write, no synchronisation required
```
Ok, done in 7afe75c3804041eebf9158686a6afd68a97a306f

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,484 @@
+        List<DiscoveryNode> getKnownPeers() {
+            assert holdsLock() : "PeerFinder mutex not held";
+            List<DiscoveryNode> knownPeers = new ArrayList<>(peersByAddress.size());
+            for(final Peer peer : peersByAddress.values()) {
```
Fixed somewhere along the way.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,484 @@
+    /**
+     * Called on receipt of a PeersResponse from a node that believes it's an active leader, which this node should therefore try and join.
+     */
+    protected abstract void onMasterFoundByProbe(DiscoveryNode masterNode, long term);
```
Ok, done in 27e95cf1675d9b7dfe9516bc0ec480ea02ab83fe

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,484 @@
+
+    public interface TransportAddressConnector {
+        /**
+         * Identify the node at the given address and, if it is a master node and not the local node then establish a full connection to it.
```
I think this is benign, but there are things we could do in the `TransportAddressConnector` to avoid this. For instance, if we are already connected to a non-master node then there's no need to identify it via a lightweight connection; perhaps we could cache responses too.

This PR is large enough without another implementation and test suite - the real implementation can come in a followup.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,484 @@
+                        @Override
+                        public void handleResponse(PeersResponse response) {
+                            logger.trace("{} received {} from {}", this, response, discoveryNode);
+                            final boolean foundMasterNode;
```
Ok, done in 5374f2b542a6db800daf09990ae8af1879469f82

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,484 @@
+                        assert remoteNode.equals(getLocalNode()) == false : remoteNode + " is the local node";
+                        synchronized (mutex) {
+                            if (running) {
+                                discoveryNode.set(remoteNode);
```
Ok, done in 1308444fa15e6b5a0f3663fc507f853571ca0e99

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,484 @@
+                assert removed == Peer.this;
+            }
+
+            private void requestPeers(final DiscoveryNode discoveryNode) {
```
Ok, done in b7c598b441c24632f6d4fea0832b12fded351165

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,484 @@
+                    protected void doRun() {
+                        // No synchronisation required for most of this
+                        List<TransportAddress> providedAddresses
+                            = new ArrayList<>(hostsProvider.buildDynamicHosts((hosts, limitPortCounts)
```
Ok, done in 417e6e26abee53f1dce82e1d263c0aa095ef02a2

DaveCTurner commented:
I merged `ActivePeerFinder` and `PeerFinder` since all it needs is a boolean saying whether it's active or not, and added assertions that when deactivated it clears out the list of known peers.

I also extracted out `ConfiguredHostsResolver` since it was only this that caused `PeerFinder` to need a lifecycle and to muck around with creating its own executor service. The actual implementation is gone, to be restored in a followup.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,378 @@
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.coordination;
```
Arguably this should be `org.elasticsearch.discovery`. WDYT?

ywelsch replied:
ok

DaveCTurner replied:
Ok, moved things around in 2013f10d14b6764a23f1c8c5a2a635a9cab3bb5c

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,383 @@
+
+public abstract class PeerFinder extends AbstractComponent {
+
+    public static final String REQUEST_PEERS_ACTION_NAME = "internal:discovery/requestpeers";
```
`request_peers`?

DaveCTurner replied:
Sure, see 1d4adccacbb835372efa564216f3a3946686a753

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,383 @@
+            (request, channel, task) -> channel.sendResponse(handlePeersRequest(request)));
+    }
+
+    public Iterable<DiscoveryNode> getFoundPeers() {
```
With the changes that introduced `handlePeersRequest` in this class, do we still need to expose this method?

DaveCTurner replied:
Yes, the `PeerFinder` has to expose the peers it's found somehow, so that we can elect one of them as a master.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,383 @@
+    public void deactivate(DiscoveryNode leader) {
+        synchronized (mutex) {
+            logger.trace("deactivating PeerFinder and setting leader to {}", leader);
+            if (active) {
```
why make this method idempotent? Is this required somewhere?

DaveCTurner replied:
Not any more, see f0e7155dec79f83ec76c71c44010c07b9d928b83

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,383 @@
+            this.transportAddress = transportAddress;
+        }
+
+        DiscoveryNode getDiscoveryNode() {
```
`@Nullable`

DaveCTurner replied:
See 4c52e503b8588d7913462e86a27d466ee2803c15

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,383 @@
+                knownPeers.add(peerNode);
+            }
+        }
+        return knownPeers;
```
maybe simpler to write `return peersByAddress.values().stream().map(Peer::getDiscoveryNode).filter(Objects::nonNull).collect(Collectors.toList());`

DaveCTurner replied:
Sure, see 121aad305986d6ce2fc01a67d72c17d0248aedf8

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,383 @@
+        }
+
+        if (active == false) {
+            logger.trace("ActivePeerFinder#handleWakeUp(): not running");
```
rename to PeerFinder

DaveCTurner replied:
Gotta catch them all. See 2051db6363d1b5105ab578b3ce390e4c655d44ee.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,383 @@
+                                if (masterNode.equals(discoveryNode) == false) {
+                                    startProbe(masterNode.getAddress());
+                                }
+                            } else {
```
I wonder if we should even make this distinction. I think the following should suffice:

```
response.getMasterNode().map(DiscoveryNode::getAddress).ifPresent(PeerFinder.this::startProbe);
response.getKnownPeers().stream().map(DiscoveryNode::getAddress).forEach(PeerFinder.this::startProbe);
```

DaveCTurner replied:
It's all about the one-liners today. See 221253a3f656a1e47d28ff30c551e33032be5a61.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,383 @@
+
+        @Override
+        public String toString() {
+            return "Peer{" + transportAddress + " peersRequestInFlight=" + peersRequestInFlight + "}";
```
also add discoveryNode?

DaveCTurner replied:
Sure, see e5414a170acd19535823c02537f6e68dd702b7a1

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,383 @@
+
+                    @Override
+                    public void handleResponse(PeersResponse response) {
+                        logger.trace("{} received {} from {}", Peer.this, response, discoveryNode);
```
with discoveryNode added to Peer.toString, no need to add discoveryNode here.

DaveCTurner replied:
Sure, see e5414a170acd19535823c02537f6e68dd702b7a1

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,383 @@
+                    @Override
+                    public void handleException(TransportException exp) {
+                        peersRequestInFlight = false;
+                        logger.debug("PeersRequest failed", exp);
```
add Peer.this to message?

DaveCTurner replied:
Sure, see e5414a170acd19535823c02537f6e68dd702b7a1

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -0,0 +1,383 @@
+            final DiscoveryNode discoveryNode = getDiscoveryNode();
+            assert discoveryNode != null : "cannot request peers without first connecting";
+
+            logger.trace("{} requesting peers from {}", this, discoveryNode);
```
no need for the `from {}` with the proposed toString change.

DaveCTurner replied:
Sure, see e5414a170acd19535823c02537f6e68dd702b7a1

ywelsch commented on server/src/main/java/org/elasticsearch/discovery/ConfiguredHostsResolver.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.List;
+import java.util.function.Consumer;
+
+public interface ConfiguredHostsResolver {
```
I'm not sure this should be a top-level class. It feels more like a helper to PeerFinder.

DaveCTurner replied:
Ok, see 8b605ada5cfcd6a76704fceb2fc38aceda86cd2a

ywelsch commented on server/src/main/java/org/elasticsearch/discovery/ConfiguredHostsResolver.java:
```diff
@@ -0,0 +1,36 @@
+     */
+    void resolveConfiguredHosts(Consumer<List<TransportAddress>> consumer);
+}
+
```
one newline too much

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -180,7 +180,7 @@ private void handleWakeUp() {
 
         if (active == false) {
-            logger.trace("ActivePeerFinder#handleWakeUp(): not running");
+            logger.trace("PeerFinder: not running");
```
do we even need to say that this is PeerFinder? We have a logger for PeerFinder that already spits out the class name (same thing for other logging occurrences in this class).

DaveCTurner replied:
Of course. I pushed 7948268c09af8bc024a1d29ee2710bd22dec6462.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java:
```diff
@@ -102,10 +102,8 @@ public void activate(final DiscoveryNodes lastAcceptedNodes) {
-                active = false;
-                handleWakeUp();
-            }
+            active = false;
```
The tests deactivate the `PeerFinder` at the end, regardless of whether it's active or not, and there's no particular reason to avoid double-deactivation or to make sure that each test ends with an active `PeerFinder`.

Also although I am pretty sure that we can't change the leader without being active, this is not something we guarantee, nor do I think we need to.
